core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/stopstartds/StopStartDownscaleActions.java

```diff
@@ -91,15 +92,34 @@ protected void prepareExecution(StopStartDownscaleTriggerEvent payload, Map<Obje
             }
 
             @Override
-            protected void doExecute(
-                    StopStartDownscaleContext context, StopStartDownscaleTriggerEvent payload, Map<Object, Object> variables) throws Exception {
-                stopStartDownscaleFlowService.clusterDownscaleStarted(context.getStack().getId(), payload.getHostGroup(), payload.getHostIds());
-                sendEvent(context);
+            protected void doExecute(StopStartDownscaleContext context, StopStartDownscaleTriggerEvent payload, Map<Object, Object> variables) throws Exception {
+                StackDtoDelegate stack = context.getStack();
+                stopStartDownscaleFlowService.clusterRecoveryCandidatesCollectionStarted(stack.getId(), payload.getHostGroup());
+                List<InstanceMetadataView> allInstanceMetadataInHostGroup =
+                        instanceMetaDataService.getAllAvailableInstanceMetadataViewsByStackId(stack.getId());
+
+                List<CloudInstance> allInstancesInHostGroup = instanceMetaDataToCloudInstanceConverter.convert(allInstanceMetadataInHostGroup,
+                        stack.getStack());
+                StopStartDownscaleGetRecoveryCandidatesRequest request = new StopStartDownscaleGetRecoveryCandidatesRequest(context.getCloudContext(),
+                        context.getCloudCredential(), context.getCloudStack(), context.getHostGroupName(), allInstancesInHostGroup, payload.getHostIds());
+                sendEvent(context, request);
             }
+        };
+    }
+
+    // TODO CB-15132: Potential pre-flight checks. YARN / appropriate service masters available. CM master available.
+    @Bean(name = "STOPSTART_DOWNSCALE_HOSTS_DECOMMISSION_STATE")
+    public Action<?, ?> decommissionViaCmAction() {
+        return new AbstractStopStartDownscaleActions<>(StopStartDownscaleGetRecoveryCandidatesResult.class) {
 
             @Override
-            protected Selectable createRequest(StopStartDownscaleContext context) {
-                return new StopStartDownscaleDecommissionViaCMRequest(context.getStack().getId(), context.getHostGroupName(), context.getHostIdsToRemove());
+            protected void doExecute(
+                    StopStartDownscaleContext context, StopStartDownscaleGetRecoveryCandidatesResult payload, Map<Object, Object> variables) throws Exception {
+                stopStartDownscaleFlowService.clusterDownscaleStarted(context.getStack().getId(), payload.getHostGroupName(), payload.getHostIds(),
```

Comment:
Haven't looked at the messaging flow in detail yet. TBD.

In the next step where instances are stopped. May be worth differentiating which ones were regular vs recovery candidates.

Reply:
Include count in messages.properties for acticity console